flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/StopWithSavepointTerminationHandlerImpl.java

```diff
@@ -0,0 +1,258 @@
+                        },
+                        mainThreadExecutor)
+                .thenCompose(
+                        aVoid ->
+                                terminatedExecutionsFuture.thenAcceptAsync(
+                                        this::handleExecutionsTermination, mainThreadExecutor));
+
+        return result;
+    }
+
+    private synchronized void handleSavepointCreation(CompletedCheckpoint completedCheckpoint) {
+        final State oldState = state;
+        state = state.onSavepointCreation(completedCheckpoint);
+
+        log.debug(
+                "Stop-with-savepoint transitioned from {} to {} on savepoint creation handling for job {}.",
+                oldState,
+                state,
+                jobId);
+    }
+
+    private synchronized void handleSavepointCreationFailure(Throwable throwable) {
+        final State oldState = state;
+        state = state.onSavepointCreationFailure(throwable);
+
+        log.debug(
+                "Stop-with-savepoint transitioned from {} to {} on savepoint creation failure handling for job {}.",
+                oldState,
+                state,
+                jobId);
+    }
+
+    private synchronized void handleExecutionsTermination(
+            Collection<ExecutionState> executionStates) {
+        final State oldState = state;
+        state = state.onExecutionsTermination(executionStates);
+
+        log.debug(
+                "Stop-with-savepoint transitioned from {} to {} on execution termination handling for job {}.",
+                oldState,
+                state,
+                jobId);
+    }
+
+    /**
+     * Handles the termination of the {@code StopWithSavepointTerminationHandler} exceptionally
+     * after triggering a global job fail-over.
+     *
+     * @param completedSavepoint the completed savepoint that needs to be discarded.
+     * @param unfinishedExecutionStates the unfinished states that caused the failure.
+     */
+    private void terminateExceptionallyWithGlobalFailover(
+            CompletedCheckpoint completedSavepoint,
+            Iterable<ExecutionState> unfinishedExecutionStates) {
+        String errorMessage =
+                String.format(
+                        "Inconsistent execution state after stopping with savepoint. At least one execution is still in one of the following states: %s. A global fail-over is triggered to recover the job %s.",
+                        StringUtils.join(unfinishedExecutionStates, ", "), jobId);
+        FlinkException inconsistentFinalStateException = new FlinkException(errorMessage);
+
+        scheduler.handleGlobalFailure(inconsistentFinalStateException);
+
+        try {
+            completedSavepoint.discard();
+        } catch (Exception e) {
+            log.warn(
+                    "Error occurred while cleaning up completed savepoint due to stop-with-savepoint failure.",
+                    e);
+            inconsistentFinalStateException.addSuppressed(e);
+        }
```

Comment:
This can entail IO operations. IO operations should not be executed in the main thread because they can be blocking. If we want to discard the savepoint, then we should pass in an `ioExecutor` which can be used for this.

Reply:
Alternatively, we need to give the `completedSavepoint` to the owner of this class in case of a termination failure. The owner would then be responsible for cleaning the savepoint up.

Reply:
👍 I left it in `StpWithSavepointTerminationHandlerImpl` because it's part of the actual handle logic and therefore is easier to be tested there.